Add logic to step through the analysis board preview with the mouse wheel

diogoaj opened the pull request:
Resolves #8738 

* Added functionality to step through the analysis board lines preview with the mouse wheel.

Not sure if this is the best way to do it, would appreciate feedback.

benediktwerner commented on ui/ceval/src/view.ts:
```diff
@@ -262,13 +262,28 @@ function getElUci(e: MouseEvent): string | undefined {
   );
 }
 
+let pvMoves: string | any[] = [];
```
This type doesn't look right? This should probably be `(string | null)[]` as well?

diogoaj replied:
Done

benediktwerner commented on ui/ceval/src/view.ts:
```diff
@@ -298,10 +313,23 @@ export function renderPvs(ctrl: ParentCtrl): VNode | undefined {
             const pvBoard = (e.target as HTMLElement).getAttribute('data-board');
             if (pvBoard) {
+              pvMoves = getElPvMoves(e);
+              pvIndex = pvMoves.indexOf(pvBoard);
```
This won't work when a position repeats, which I'm pretty sure can happen in Stockfish lines.

diogoaj replied:
I did not test with a specific example but I think it works because ```pvMoves = getElPvMoves(e);``` fetches the moves from the line that is hovered.  ```pvMoves.indexOf(pvBoard);``` gets me the index of the current move so I can go back and forth inside that line.

diogoaj replied:
Yeah I get what you are saying, I have to look into this with a specific example.

diogoaj replied:
I think my last commit fixes the issue. It ensures we get the correct move index.

benediktwerner commented on ui/ceval/src/view.ts:
```diff
@@ -298,10 +313,23 @@ export function renderPvs(ctrl: ParentCtrl): VNode | undefined {
+            if (e.deltaY < 0 && pvIndex > 0) pvIndex -= 1; 
+            else if (e.deltaY > 0 && pvIndex < pvMoves.length) pvIndex += 1;
+
+            const pvBoard = pvMoves[pvIndex]!;
```
You shouldn't need a `!` here

diogoaj replied:
Done

benediktwerner commented:
Also, your code isn't quite formatted correctly. You can use Prettier to auto-format it, either in your editor or via `yarn run format` at the root of the project.

ornicar commented:
all this code and global state scares me, it looks like a potential source of future bugs and maintenance work.

Also doesn't the index need to be reset, like when the variations change?

diogoaj commented:
> all this code and global state scares me, it looks like a potential source of future bugs and maintenance work.
>
> Also doesn't the index need to be reset, like when the variations change?

Made the necessary changes in my last commit 8ffdc7a, hope it addresses these issues.

benediktwerner commented:
`move-index` should be a data-attribute (i.e. `data-move-index`). You can also use `el.dataset["move-index"]` and `$el.data("move-index")` instead of `el.getAttribute("data-move-index")` and `$el.attr("data-move-index")`.

diogoaj commented:
> `move-index` should be a data-attribute (i.e. `data-move-index`). You can also use `el.dataset["move-index"]` and `$el.data("move-index")` instead of `el.getAttribute("data-move-index")` and `$el.attr("data-move-index")`.

I see, did not know about that. Thanks!

ornicar commented on ui/ceval/src/view.ts:
```diff
@@ -379,6 +414,7 @@ function renderPvMoves(pos: Position, pv: Uci[]): VNode[] {
         {
           key,
           attrs: {
+            'data-move-index': i,
```
does that work with `.dataset['moveIndex']`?

diogoaj replied:
Yeah, I didn't know either, but the dashes are converted to camel case when accessing the data.
